src/analysis/Environments/Environments.ts

```diff
@@ -238,29 +206,37 @@ export function PauseButton({ cloudEnvironment, currentUser, pauseComputeAndRefr
     : null;
 }
 
-interface DecoratedResourceAttributes {
-  workspace: WorkspaceInfo;
-  unsupportedWorkspace: boolean;
+export interface UseWorkspacesResult {
+  workspaces: WorkspaceWrapper[];
+  refresh: () => Promise<void>;
+  loading: boolean;
 }
 
-type RuntimeWithWorkspace = DecoratedResourceAttributes & Runtime;
-type DiskWithWorkspace = DecoratedResourceAttributes & PersistentDisk;
-type AppWithWorkspace = DecoratedResourceAttributes & App;
+export type UseWorkspacesProvider = (
+  fields?: Record<string, string>,
+  stringAttributeMaxLength?: string | number
+) => UseWorkspacesResult;
 
-type DecoratedComputeResource = RuntimeWithWorkspace | AppWithWorkspace;
-type DecoratedResource = DecoratedComputeResource | DiskWithWorkspace;
+type LeoAppProviderNeeds = Pick<LeoAppProvider, 'listWithoutProject' | 'pause' | 'delete'>;
+type LeoRuntimeProviderNeeds = Pick<LeoRuntimeProvider, 'list' | 'stop' | 'delete'>;
+type LeoDiskProviderNeeds = Pick<LeoDiskProvider, 'list' | 'delete'>;
 
 export interface EnvironmentsProps {
   nav: NavLinkProvider<EnvironmentNavActions>;
+  useWorkspacesState: UseWorkspacesProvider;
+  leoAppData: LeoAppProviderNeeds;
+  leoRuntimeData: LeoRuntimeProviderNeeds;
+  leoDiskData: LeoDiskProviderNeeds;
+  metrics: MetricsProvider;
 }
 
-export const Environments: React.FC<EnvironmentsProps> = (props) => {
-  const { nav } = props;
+export const Environments = (props: EnvironmentsProps): ReactNode => {
+  const { nav, useWorkspacesState, leoAppData, leoDiskData, leoRuntimeData, metrics } = props;
   const signal = useCancellation();
   const { workspaces, refresh: refreshWorkspaces } = _.flow(
-    useWorkspaces,
+    useWorkspacesState,
     _.update('workspaces', _.flow(_.groupBy('workspace.namespace'), _.mapValues(_.keyBy('workspace.name'))))
-  )();
+  )() as UseWorkspacesResult;
```

Comment:
This is not a `UseWorkspacesResult`. The `_.update` function transforms the list of workspaces into an object of:
```
{
  [billingProject]: {
    [workspaceName]: workspace,
    ...
  },
  ...
}
```

Suggest breaking this up into two statements to clarify the transformation.

Reply:
Yeah. I think they're best used sparingly because of that. But I think the power of being able to do complex transformations in a single line can be worth it. As long as the result is stored in a typed variable to contain the spread of `any`.

```ts
const { workspaces: listWorkspaces } = useWorkspaces();
const workspaces: { [namespace: string]: { [name: string]: WorkspaceWrapper } } = _.flow(..., workspacesList);
```

Reply:
fixed with new Mutate utility type.... check it out...